tensorflow/tsl/platform/cpu_info.cc

```diff
@@ -345,6 +352,115 @@ void InitCPUIDInfo() {
 
 #endif  // PLATFORM_IS_X86
 
+#ifdef PLATFORM_IS_ARM64
+
+class CPUIDInfo;
+void InitCPUIDInfo();
+
+CPUIDInfo *cpuid = nullptr;
+
+// Structure for basic CPUID info
+class CPUIDInfo {
+ public:
+  CPUIDInfo()
+      : implementer_(0),
+        variant_(0),
+        cpunum_(0),
+        is_arm_neoverse_v1_(0),
+        is_arm_neoverse_n1_(0) {}
+
+  static void Initialize() {
+    // Initialize cpuid struct.
+    if (cpuid != nullptr) {
+      return;
+    }
+    cpuid = new CPUIDInfo;
+
+    if (!(getauxval(AT_HWCAP) & HWCAP_CPUID)) {
+      return;
+    }
+
+    std::ifstream CPUspresent;
+    CPUspresent.open("/sys/devices/system/cpu/present", std::ios::in);
+    int present_cpu = -1;
+    if (CPUspresent.is_open()) {
+      std::string line;
+      if (bool(getline(CPUspresent, line))) {
+        // We just need to find one CPU that is active
+        // from which we can read MIDR register to find
+        // implement, variant and revision information.
+        auto ending = line.end();
+        for (auto i = line.begin(); i < line.end(); ++i) {
+          if (*i == '-' || *i == ',') {
+            ending = i;
+            break;
+          }
+        }
+        line.erase(ending, line.end());
+        // That should be the fist number.
+        present_cpu = std::stoi(line);
+      }
+    }
+
+    if (present_cpu == -1) {
+      return;
+    }
+
+    std::stringstream str;
+    str << "/sys/devices/system/cpu/cpu" << present_cpu
+        << "/regs/identification/midr_el1";
+    std::ifstream midr_el1_file(str.str(), std::ios::in);
+    if (midr_el1_file.is_open()) {
+      std::string line;
+      if (bool(getline(midr_el1_file, line))) {
+        uint32 midr_el1 = std::stoul(line, nullptr, 16);
+
+        // Unpack variant and CPU ID
+        cpuid->implementer_ = (midr_el1 >> 24) & 0xFF;
+        cpuid->variant_ = (midr_el1 >> 20) & 0xF;
+        cpuid->cpunum_ = (midr_el1 >> 4) & 0xFFF;
+        if (cpuid->implementer_ == 0x41) {
+          switch (cpuid->cpunum_) {
+            case 0xd40:  // ARM NEOVERSE V1
+              cpuid->is_arm_neoverse_v1_ = 1;
+            case 0xd0c:  // ARM NEOVERSE N1
+              cpuid->is_arm_neoverse_n1_ = 1;
+            default:
+              break;
+          }
+        }
+      }
+    }
+  }
+
+  int implementer() const { return implementer_; }
+  int cpunum() const { return cpunum_; }
+
+  static bool TestAarch64CPU(Aarch64_CPU cpu) {
+    InitCPUIDInfo();
+    // clang-format off
+    switch (cpu) {
+      case ARM_NEOVERSE_V1: return cpuid->is_arm_neoverse_v1_;
+      default:
+        return 0;
+        break;
```

Comment:
I don't think we need this line since we just had a `return 0;`.

Reply:
Done